### DISCUSSION TITLE
iPXE not honoring options 66 & 67 from dhcp server

### DISCUSSION THREAD

**Discussion by nod0ubt** [ipxe/ipxe#209] [Q&A] [answered]
I've got a strange issue.
Our Citrix XenServer hypervisor utilizes iPXE. I do have few VMs which I'm using for testing purposes.
There's isc-dhcp server with "next-server" option pointing to FOG server and that works as intended. However, recently I started using the built-in dhcp server of our core switch. There I'm only able to set options in numerical value (no "next-server"). So I configured option 66 and 67 with the ip of the server and boot file name, but it seems that iPXE doesn't like that and won't boot.

After receiving IP from dhcp I'm getting this error:

```
Filename: undionly.kpxe
Could not start download: Operation not supported (http://ipxe.org/3c092003)
```

I have inspected the frames with Wireshark and it seems that iPXE only boots when I have "next-server" and "Boot file name" in the frame body. When there's Option 66 and Option 67 it errors with the mentioned above error.
Am I doing something wrong? Please advice.

**Comment by mcb30** [comment, marked as answer]
As per https://tools.ietf.org/html/rfc2132#section-9.4 DHCP option 66 is "TFTP server name" and is equivalent to the BOOTP `sname` field.  This is not the TFTP server address, which is instead carried in the `siaddr` field.  The definition of `siaddr` and `sname` is given in https://tools.ietf.org/html/rfc2131#section-2

In summary: option 66 is not supposed to be used to convey the TFTP server IP address, and iPXE is correctly ignoring it.

If you are not able to configure the switch to provide the TFTP server address via the `siaddr` field (as required by the RFC) then you could work around the problem by specifying a URL as the boot file name in option 67.

For example, if you set option 67 to `http://boot.ipxe.org/demo/boot.php` then iPXE will boot directly via HTTP with no TFTP stage required.  You can also, if you wish, set a TFTP URL such as `tftp://192.168.0.1/wdsnbp.com`

**Reply by nod0ubt** [reply]
Spot on! Thanks, Michael.
Your instructions really helped me.
The boot via HTTP options works nicely. However, the TFTP stage seems to fail somewhere as I'm getting "Please enter tftp server:" message.
Manually entering the IP works. Not sure if that's an issue with iPXE though.

**Reply by mcb30** [reply]
> The boot via HTTP options works nicely. However, the TFTP stage seems to fail somewhere as I'm getting "Please enter tftp server:" message.
> Manually entering the IP works. Not sure if that's an issue with iPXE though.

The string "Please enter tftp server" is not something that appears anywhere in the iPXE code, so that must be coming from something else.  Do you have a screenshot?

**Reply by nod0ubt** [reply]
Yes, Michael, it probably is comming from the FOG server.
![image](https://user-images.githubusercontent.com/37583437/104640322-00f21f80-56b1-11eb-8442-acf3cf57e052.png)

**Reply by mcb30** [reply]
That looks like a prompt produced by an iPXE script, which is presumably a script provided by FOG.

**Reply by nod0ubt** [reply]
Alright. Guess I'll talk a bit more to the FOG guys.
